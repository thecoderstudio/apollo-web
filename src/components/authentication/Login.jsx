import React from 'react';
import { Formik } from 'formik';
import axios from 'axios';
import styled from 'styled-components';
import { connect } from 'react-redux';
import Button from '../buttons/Button';
import { StatusCodes } from 'http-status-codes';
import Input from '../Input';
import loginSchema from '../../validation/login';
import { login as loginAction } from '../../actions/auth';
import { parseHTTPErrors } from '../../util/parser';
import { handleHTTPResponse } from '../../actions/error';
import { fetchCurrentUser } from '../../util/user';

const Form = styled.form`
  height: 75%;
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 40px;
`;

class Login extends React.PureComponent {
  constructor(props) {
    super(props);
    this.login = this.login.bind(this);
    this.loginSuccessCallback = this.loginSuccessCallback.bind(this);
  }

  login(credentials, { setErrors }) {
    console.log("login")
    axios.post(
      `${process.env.APOLLO_HTTP_URL}auth/login`,
      credentials,
      { withCredentials: true }
    )
      .then(_ => {
        console.log("SUCCES")
        fetchCurrentUser(this.loginSuccessCallback);
      })
      .catch(error => {
        handleHTTPResponse(error.response, true, true);
        if (error.response.status === StatusCodes.BAD_REQUEST) {
          setErrors(parseHTTPErrors(error.response.data, { detail: 'password' }));
        }
      });
  }

  loginSuccessCallback() {
    this.props.dispatch(loginAction());
  }

  render() {
    return (
      <Formik
        initialValues={{ username: '', password: '' }}
        validationSchema={loginSchema}
        validateOnChange={false}
        onSubmit={this.login}>
        {({ values, errors, handleChange, handleSubmit }) => (
          <Form onSubmit={handleSubmit}>
            <Input
              name="username"
              type="username"
              placeholder="Username"
              value={values.username}
              error={errors.username}
              onChange={handleChange} />
            <Input
              name="password"
              type="password"
              placeholder="Password"
              value={values.password}
              error={errors.password}
              onChange={handleChange} />
            <Button>Log in</Button>
          </Form>
        )}
      </Formik>
    );
  }
}

export default connect()(Login);

export { Login as UnconnectedLogin };
