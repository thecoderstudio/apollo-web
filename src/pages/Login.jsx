import React from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import styled from 'styled-components';
import Button from '../components/Button';
import Input from '../components/Input';
import Card from '../components/Card';
import { login as loginAction } from '../actions/auth';
import media from '../util/media';
import moonImg from '../images/moon_rocket.svg';

const OuterContainer = styled.div`
  height: 100%;
  display: grid;
  align-items: center;
`;

const InnerContainer = styled.div`
  display: grid;
  justify-items: center;
  column-gap: 50px;
  height: 90%;
  max-height: 650px;
  margin: 100px;
  grid-template-rows: [heading] 1fr [title] 1fr [content] 4fr;
  grid-template-columns: [img] 1fr [content] 1fr;

  ${
    media.phone`
      grid-template-columns: [content] 1fr;
      grid-template-rows: [title] 1fr [content] 4fr [footer] 1fr;
      margin: 20px 20px 0px 20px;
    `
  }
`;

const Title = styled.h1`
  grid-row: title;
  grid-column: content;
  color: ${props => props.theme.primary};
`;

const StyledCard = styled(Card)`
  width: 100%;
  max-width: 400px;
  height: 350px;
  grid-row: content;
  grid-column: content;
  display: grid;
  align-items: center;

  ${
    media.phone`
      box-shadow: none;
      background-color: ${props => props.theme.black};
    `
  }
`;

const Form = styled.form`
  height: 75%;
  display: grid;
  grid-template-rows: auto;
  grid-row-gap: 40px;
`;

const SupportingImg = styled.img`
  grid-column: img;
  grid-row: 2 / 4;
  width: 75%;
  max-width: 400px;
  align-self: center;

  ${
    media.phone`
      grid-column: 1;
      grid-row: footer;
      width: 100%;
    `
  }
`;

class Login extends React.Component {
  constructor(props) {
    super(props);
    this.handleUsernameChange = this.handleUsernameChange.bind(this);
    this.handlePasswordChange = this.handlePasswordChange.bind(this);
    this.login = this.login.bind(this);
    this.state = { username: '', password: '' };
  }

  handleUsernameChange(e) {
    this.setState({ username: e.target.value });
  }

  handlePasswordChange(e) {
    this.setState({ password: e.target.value });
  }

  login(e) {
    e.preventDefault();

    const credentials = {
      username: this.state.username,
      password: this.state.password
    };

    axios.post(
      process.env.APOLLO_URL.concat("auth/login"),
      credentials,
      { withCredentials: true }
    )
      .then(res => {
        switch (res.status) {
          case 200:
            this.props.dispatch(loginAction());
          default:
            return;
        };
      })
      .catch(error => {
        console.log(error)
      });
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (this.props.authenticated) {
      window.location.pathname = "/";
    }
  }

  render() {
    return (
      <OuterContainer>
        <InnerContainer>
          <SupportingImg src={moonImg} />
          <Title>Log in to Apollo</Title>
          <StyledCard>
            <Form onSubmit={this.login}>
              <Input type="username" placeholder="Username" onChange={this.handleUsernameChange} />
              <Input type="password" placeholder="Password" onChange={this.handlePasswordChange} />
              <Button>Log in</Button>
            </Form>
          </StyledCard>
        </InnerContainer>
      </OuterContainer>
    );
  }
}

export default connect(({ auth }) => ({ authenticated: auth.authenticated }))(Login);
