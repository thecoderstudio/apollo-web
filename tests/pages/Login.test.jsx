import React from 'react';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import axios from 'axios';
import waitForExpect from 'wait-for-expect';
import Login from '../../src/pages/Login';

const mockStore = configureStore([]);
jest.mock('axios');

function getComponent(store) {
  return renderer.create(
    <Provider store={store}>
      <Login />
    </Provider>
  );
}

describe('login', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      authenticated: false
    });
    store.dispatch = jest.fn();
  });

  it("renders correctly", () => {
    const tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("handles successful login", async () => {
    const component = getComponent(store);
    const root = component.root.findByProps({ authenticated: false });
    const instance = root.instance;

    axios.post.mockResolvedValue({
      status: 200
    });
    axios.get.mockResolvedValue({
      status: 200
    });

    root.findByProps({ type: 'username' }).props.onChange({
      currentTarget: {
        name: 'username',
        value: 'test'
      }
    });
    root.findByProps({ type: 'password' }).props.onChange({
      currentTarget: {
        name: 'password',
        value: 'password'
      }
    });
    root.findByType('form').props.onSubmit();

    await waitForExpect(() => {
      expect(instance.props.dispatch).toHaveBeenCalled();
    });
  });

  it("handles unsuccessful login", async () => {
    const component = getComponent(store);
    const root = component.root.findByProps({ authenticated: false });
    const instance = root.instance;

    axios.post.mockImplementationOnce(() => Promise.reject({
      response: {
        status: 400,
        statusText: "Bad request",
        data: {}
      }
    }));

    root.findByProps({ type: 'username' }).props.onChange({
      currentTarget: {
        name: 'username',
        value: 'test'
      }
    });
    root.findByProps({ type: 'password' }).props.onChange({
      currentTarget: {
        name: 'password',
        value: 'password'
      }
    });
    root.findByType('form').props.onSubmit();

    await waitForExpect(() => {
      expect(instance.props.dispatch).not.toHaveBeenCalled();
    });
  });

  it("handler get current user failure", async () => {
    const component = getComponent(store);
    const root = component.root.findByProps({authenticated: false});
    const instance = root.instance;

    axios.post.mockResolvedValue({
      status: 200
    });
    axios.get.mockImplementationOnce(() => Promise.reject({
      response: {
        status: 500,
        statusText: "Something went wrong",
        data: {}
      }
    }));


    root.findByProps({type: 'username'}).props.onChange({ 
      currentTarget: {
        name: 'username',
        value: 'test'
      }
    });
    root.findByProps({type: 'password'}).props.onChange({
      currentTarget: {
        name: 'password',
        value: 'password'
      }
    });
    root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });

    await waitForExpect(() => {
      expect(instance.props.dispatch).toHaveBeenCalledTimes(1);
    });
  });

  it("correctly sets pathname on authentication", () => {
    const component = getComponent(store);
    const root = component.root.findByProps({ authenticated: false });
    const instance = root.instance;

    instance.props = ({ authenticated: true });
    instance.componentDidUpdate();
    expect(window.location.pathname).toEqual("/");
  });
});
