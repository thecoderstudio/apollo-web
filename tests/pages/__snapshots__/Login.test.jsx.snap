// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login renders correctly 1`] = `
<div
  className="sc-AxhCb KPxTP"
>
  <div
    className="sc-AxhUy ipnyuM"
  >
    <img
      className="sc-AxmLO AMLso"
      src={[Function]}
    />
    <h1
      className="sc-AxgMl dVtKuT"
    >
      Log in to Apollo
    </h1>
    <div
      className="sc-AxiKw sc-AxheI gbjjdj"
    >
      <form
        className="sc-Axmtr hUAPTi"
        onSubmit={[Function]}
      >
        <input
          className="sc-AxirZ foZgfu"
          onChange={[Function]}
          placeholder="Username"
          type="username"
        />
        <input
          className="sc-AxirZ foZgfu"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
        <button
          className="sc-AxjAm bjwrOj"
        >
          Log in
        </button>
      </form>
    </div>
  </div>
</div>
`;
